### PR TITLE
[ROCM] Enable basic functionality to run MLIR fussion tests

### PR DESCRIPTION
**Pull request by draganmladjenovic**
Specific tests will be fixed as we go

### CODE REVIEW

File: third_party/llvm/rocdl_shuffle_down.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+     auto int32Type = IntegerType::get(rewriter.getContext(), 32);
+     Value width = adaptor.getWidth();
+     Value zero = rewriter.create<LLVM::ConstantOp>(loc, int32Type, 0);
+     Value negwidth = rewriter.create<LLVM::SubOp>(loc, int32Type, zero, width);
+     Value add = rewriter.create<LLVM::AddOp>(loc, int32Type, srcLaneId, width);
+@@ -151,6 +151,10 @@ struct GPUShuffleOpLowering : public ConvertOpToLLVMPattern<gpu::ShuffleOp> {
+     // TODO: Use ds_swizzle for XOR when step/offsets are constants for better
+     // perf.
+     switch (op.getMode()) {
++    case gpu::ShuffleMode::DOWN:
```
Comment (jreiffers) [inline third_party/llvm/rocdl_shuffle_down.patch:26]: Can you link to an LLVM PR for this?

File: xla/service/gpu/fusions/mlir/elemental_hlo_to_mlir.cc [hunk 2]
```diff
@@ -1180,20 +1180,13 @@ absl::StatusOr<SmallVector<Value, 1>> HloToMlir(
 }  // namespace
 
 bool IsHloOpSupported(const HloInstruction* instr,
```
Comment (jreiffers) [inline xla/service/gpu/fusions/mlir/elemental_hlo_to_mlir.cc:1182]: We can remove the compute capability here now (well, technically, we could have removed it before, but now we really can :-)).

File: xla/service/gpu/gpu_device_info_for_tests.cc [hunk 3]
```diff
@@ -64,5 +64,13 @@ stream_executor::DeviceDescription TestGpuDeviceInfo::AMDMI210DeviceInfo() {
   return b.BuildObject();
 }
 
+stream_executor::DeviceDescription TestGpuDeviceInfo::TestCudaOrRocmDeviceInfo() {
+#if !defined(TENSORFLOW_USE_ROCM)
```
Comment (jreiffers) [inline xla/service/gpu/gpu_device_info_for_tests.cc:68]: Nit: let's flip this condition around, so we don't do if(!x) else?

```
#if defined(TENSORFLOW_USE_ROCM)
  return AMDMI210DeviceInfo();
#else
  return RTXA6000DeviceInfo()
#endif
```

File: xla/stream_executor/rocm/rocm_platform.h [hunk 4]
```diff
@@ -63,7 +63,7 @@ class ROCmPlatform : public Platform {
   // looking in or storing to the Platform's executor cache.
   // Ownership IS transferred to the caller.
   absl::StatusOr<std::unique_ptr<StreamExecutor>> GetUncachedExecutor(
-      const StreamExecutorConfig& config) override;
```
Comment (jreiffers) [inline xla/stream_executor/rocm/rocm_platform.h:66]: Is this related?

**Review by jreiffers — APPROVED** [review]

File: xla/stream_executor/rocm/rocm_platform.h (hunk 4 above)
Comment (jreiffers) [reply xla/stream_executor/rocm/rocm_platform.h:66]: Oh, it's a fix for an unrelated breakage. I see.

Comment (dimitar-asenov) [reply xla/stream_executor/rocm/rocm_platform.h:66]: FYI, this fix is being submitted separately in https://github.com/openxla/xla/pull/15726.

File: xla/service/gpu/fusions/mlir/BUILD [hunk 5]
```diff
@@ -330,6 +334,7 @@ cc_library(
         "@llvm-project//mlir:FuncToLLVM",
         "@llvm-project//mlir:GPUDialect",
         "@llvm-project//mlir:GPUToNVVMTransforms",
+        "@llvm-project//mlir:GPUToROCDLTransforms",
         "@llvm-project//mlir:IR",
         "@llvm-project//mlir:LLVMCommonConversion",
         "@llvm-project//mlir:LLVMDialect",
```
Comment (jreiffers) [inline xla/service/gpu/fusions/mlir/BUILD:340]: Looks like we need a dependency on ROCDLDialect here.

**Comment by penpornk** [comment]
@draganmladjenovic Could you please help address @jreiffers's comment [here](https://github.com/openxla/xla/pull/15671#discussion_r1703891539)?

File: xla/service/gpu/fusions/mlir/elemental_hlo_to_mlir.cc (hunk 2 above)
Comment (jreiffers) [reply xla/service/gpu/fusions/mlir/elemental_hlo_to_mlir.cc:1182]: FYI I did this in #16085.

**Review by jreiffers — APPROVED** [review]
